### PR TITLE
Add import-spacing rule

### PR DESCRIPTION
**Pull request by mohsen1**
This PR adds `import-spacing` rule which enforces proper spacing in import declarations. 

This is implementation of https://github.com/palantir/tslint/issues/1584

#### PR checklist

- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

### CODE REVIEW

**Comment by mohsen1** [comment]
I don't see any failures in the output but my tests are failing. Can you help?

**Comment by IllusionMH** [comment]
@mohsen1 I'm unable to create PR to your fork.

You can find changes that allowed me to fix test in commit https://github.com/palantir/tslint/commit/f37c74dc84e0cf136e266936e3b7cefeae413fc5

**Comment by mohsen1** [comment]
@IllusionMH thanks for looking into it. 
What `nil` means in there?

**Comment by JKillian** [comment]
`~nil` means that the error starts there but has zero width. In this case, it probably means that the error indicates that it starts after the last character of the line (where the line break character would be I suppose). In general it's nicer to avoid outputting errors like this if possible.

**Comment by IllusionMH** [comment]
@mohsen1 `~nil` is described in [lines.ts](https://github.com/palantir/tslint/blob/master/src/test/lines.ts#L85-L86)
In this case error starts on next character after `}` (in second case) and this character is line break - which doesn't have length and can't be marked with regular `~`.

**Comment by mohsen1** [comment]
@JKillian I think this is a valid use case

File: src/rules/importSpacingRule.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        ruleName: "import-spacing",
+        description: "Ensures proper spacing between import statement keywords",
+        optionsDescription: "",
+        options: {},
+        optionExamples: [],
+        type: "style",
+        typescriptOnly: false,
+    };
+
+    public static ADD_SPACE_AFTER_IMPORT = "Add space after import";
```
Comment (mohsen1) [inline src/rules/importSpacingRule.ts]: Please feel free to correct my error wordings here.

File: src/rules/importSpacingRule.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    public static NO_LINE_BREAKS = "No line breaks are allowed in import declaration";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        const comparisonWalker = new ImportStatementWalker(sourceFile, this.getOptions());
+        return this.applyWithWalker(comparisonWalker);
+    }
+}
+
+class ImportStatementWalker extends Lint.RuleWalker {
+    private static IMPORT_KEYWORD_LENGTH = 6; // "import".length;
```
Comment (nchen63) [inline src/rules/importSpacingRule.ts]: you can just define it as `"import".length` and omit the comment. It is probably even more clear to not define this variable and just just `"import".length` everywhere

File: src/rules/importSpacingRule.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        super.visitNamespaceImport(node);
+    }
+
+    private checkModuleWithSideEffect(node: ts.ImportDeclaration) {
+        const moduleSpecifierStart = node.moduleSpecifier.getStart();
+        const nodeStart = node.getStart();
+        if ((nodeStart + ImportStatementWalker.IMPORT_KEYWORD_LENGTH + 1) < moduleSpecifierStart) {
+            this.addFailure(this.createFailure(nodeStart, moduleSpecifierStart - nodeStart, Rule.TOO_MANY_SPACES_AFTER_IMPORT));
+        }
+        if ((nodeStart + ImportStatementWalker.IMPORT_KEYWORD_LENGTH) === moduleSpecifierStart) {
```
Comment (nchen63) [inline src/rules/importSpacingRule.ts]: `else if`

File: src/rules/importSpacingRule.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            } else if (importClauseStart > (importKeywordEnd + 1)) {
+                this.addFailure(this.createFailure(nodeStart, importClauseStart - nodeStart, Rule.TOO_MANY_SPACES_AFTER_IMPORT));
+            }
+
+            const fromString = node.getText().substring(importClauseEnd - nodeStart, moduleSpecifierStart - nodeStart);
+
+            if (/from$/.test(fromString)) {
+                this.addFailure(this.createFailure(importClauseEnd, fromString.length, Rule.ADD_SPACE_AFTER_FROM));
+            }
+            if (/from\s{2,}$/.test(fromString)) {
```
Comment (nchen63) [inline src/rules/importSpacingRule.ts]: `else if`

File: src/rules/importSpacingRule.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                this.addFailure(this.createFailure(importClauseEnd, fromString.length, Rule.TOO_MANY_SPACES_AFTER_FROM));
+            }
+            if (/^\s{2,}from/.test(fromString)) {
+                this.addFailure(this.createFailure(importClauseEnd, fromString.length, Rule.TOO_MANY_SPACES_BEFORE_FROM));
+            }
+            if (/^from/.test(fromString)) {
```
Comment (nchen63) [inline src/rules/importSpacingRule.ts]: `else if`

File: src/rules/importSpacingRule.ts (hunk 2 above)
Comment (mohsen1) [inline src/rules/importSpacingRule.ts]: Isn't it more performance friendly to do this?

Comment (nchen63) [inline src/rules/importSpacingRule.ts]: The perf difference is not measurable. See the jsperf run https://jsperf.com/str-length

**Comment by nchen63** [comment]
@mohsen1 thanks!

**Comment by adidahiya** [comment]
The banning of line breaks in import statements is problematic here. We need to prescribe a way to make this rule play nicely with max-line-length out of the box. Right now we don't have an option for the line length rule to skip import statements (https://github.com/palantir/tslint/issues/1872#issuecomment-267085885). IMO I would prefer to _keep_ the max-line-length rule in effect for import statements and allow multiline import statements in the default TSLint config.

```ts
import {
    RuleWalker,
    RuleFailure,
    ...
} from "tslint";
```

**Comment by mohsen1** [comment]
It does allow line breaks in destructuring import clause. although it won't let a very long module specifier like this:

```ts
import * as veryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongName from './very-very-very-very-very-very-very-very-very-long-name';
``` 

But a line break might not solve the issue if the specifier itself is longer than max lenght.

**Comment by adidahiya** [comment]
Ah, cool, sorry I missed that. This implementation looks reasonable then.

**Pull request merged**